python/ribasim/ribasim/model.py

```diff
@@ -280,6 +285,137 @@ def write(self, filepath: str | PathLike[str]) -> Path:
+
+        # filter graph by edge type
+        df_graph = df_graph.loc[df_graph["edge_type"] == edge_type]
+
+        # count occurrence of "from_node" which reflects the number of outneighbors
+        from_node_count = (
+            df_graph.groupby("from_node_id").size().reset_index(name="from_node_count")  # type: ignore
+        )
+
+        # append from_node_count column to from_node_id and from_node_type
+        from_node_info = (
+            df_graph[["from_node_id", "from_node_type"]]
+            .drop_duplicates()
+            .merge(from_node_count, on="from_node_id", how="left")
+        )
+        from_node_info = from_node_info[
+            ["from_node_id", "from_node_count", "from_node_type"]
+        ]
+
+        # add the node that is not the upstream of any other nodes
+        from_node_info = self._add_source_sink_node(nodes, from_node_info, "from")
+
+        # loop over all the "from_node" and check if they have enough outneighbor
+        for _, row in from_node_info.iterrows():
+            # from node's outneighbor
+            if row["from_node_count"] < edge_amount[row["from_node_type"]][2]:
+                is_valid = False
+                logging.error(
+                    f"Node {row['from_node_id']} must have at least {edge_amount[row['from_node_type']][2]} outneighbor(s) (got {row['from_node_count']})"
+                )
+
+        # count occurrence of "to_node" which reflects the number of inneighbors
+        to_node_count = (
+            df_graph.groupby("to_node_id").size().reset_index(name="to_node_count")  # type: ignore
+        )
+
+        # append to_node_count column to result
+        to_node_info = (
+            df_graph[["to_node_id", "to_node_type"]]
+            .drop_duplicates()
+            .merge(to_node_count, on="to_node_id", how="left")
+        )
+        to_node_info = to_node_info[["to_node_id", "to_node_count", "to_node_type"]]
+
+        # add the node that is not the downstream of any other nodes
+        to_node_info = self._add_source_sink_node(nodes, to_node_info, "to")
+
+        # loop over all the "to_node" and check if they have enough inneighbor
+        for _, row in to_node_info.iterrows():
+            if row["to_node_count"] < edge_amount[row["to_node_type"]][0]:
+                is_valid = False
+                logging.error(
+                    f"Node {row['to_node_id']} must have at least {edge_amount[row['to_node_type']][0]} inneighbor(s) (got {row['to_node_count']})"
+                )
+
+        return is_valid
+
+    def _add_source_sink_node(
+        self, nodes, node_info: pd.DataFrame, direction: str
+    ) -> pd.DataFrame:
+        """Loop over node table.
+
+        Add the nodes whose id are missing in the from_node and to_node column in the edge table because they are not the upstream or downstrem of other nodes.
+
+        Specify that their occurrence in from_node table or to_node table is 0.
+        """
+
+        if direction == "from":
```

Comment:
You could create a column variable `f"{direction}_node_id"`, and use it to remove the if/else.